ovlfs-2.0.1/doc/ovlmount.8
.\"
.\" COPYRIGHT (C) 2003 By Arthur Naseef
.\"
.\" This file is covered under the GNU General Public License Version 2.  For
.\"  more information, see the file COPYING.
.\"
.\"
.\" FILE: ovlfsmount.m
.\"
.\" DESCRIPTION:
.\"	Manual page for the ovlfsmount program which is designed to manage
.\"	ovlfs mounts more easily than fstab.
.\"
.\"
.\" REVISION HISTORY
.\"
.\" DATE	AUTHOR		DESCRIPTION
.\" ===========	===============	==============================================
.\" 2003-06-01	ARTHUR N.	Initial Release
.\"
.\"
.\" M4 definitions:
.\" COPYRIGHT (C) 1998 By Arthur Naseef
.\"
.\" This file is covered under the GNU General Public License Version 2.  For
.\"  more information, see the file COPYING.
.\"
.\"
.\" FILE: tmac.asn
.\"
.\" DESCRIPTION: This file contains general purpose nroff/troff macros.
.\"
.\" SCCS ID: @(#) tmac.asn 1.1
.\"
.\"
.\" REVISION HISTORY
.\"
.\" DATE	AUTHOR		DESCRIPTION
.\" ===========	===============	==============================================
.\" 03/08/1998	ARTHUR N.	Initial Release
.\"
.\"
.\" **************************************************************************
.\"
.\" START OF LIST MACROS
.\"
.de BL    \" Begin Bullet List
.in +8n             \" increase indent by about 8 characters
.nr *i \\n(.i/1n    \" remember the current indent in units of 'en'
.nr *i +2           \" add two more characters to the list item indent
.ds OL \\*(LM       \" remember the last mark - can only nest 2 levels
.ds LM \(bu         \" set the list item marker to a bullet
..
.de LE    \" End List
.nr *i -10      \" calculate original indent
.in \\n(*in     \" reset original indent
.ds LM \\*(OL   \" restore last marker saved
..
.de LI    \" List Item
.in \\n(*in    \" indent to list item space
.br            \" make sure to start a new line
.ti -2n        \" indent, next line only, back about 2 characters
\\*(LM         \" display list marker
..
.\"
.\" **************************************************************************
.\"
.\" START OF THE HANGING INDENT MACRO
.\"
.\" begin hanging indent from current point after the text of argument 1,
.\"  if defined.
.\"
.nr %I 0
.nr w 1n	\" Set this once and allow it to be changed later
.de HI
.if \\n(%I-9  .ab HI macro stack error    \" If %I > 9, can't continue
.nr I\\n(%I \\n(.iu
.nr %I \\n(%I+1
.di *H           \" Divert output to *H macro and put $1 on input
\&\\$1
.br              \" Break this input to put in output (so the diversion gets it)
.di              \" End the diversion
.nr $i \\n(dlu+\\nwu  \" Calculate the hanging indentation
.nr %i \\nw-\w'  '    \" Take offset and subtract the spaces that are coming
.ti 0
\&\\*(*H\\h'\\n(%iu'
'in \\n($iu
..
.\"
.\" end of hanging indent - restore the PREVIOUS indent
.\"
.de HE
.nr %I \\n(%I-1       \" Decrement current indentation counter
.in \\n(I\\n(%Iu      \" Restore indentation at current counter
..
.\"
.\" END OF THE HANGING INDENT MACRO
.\"
.\" **************************************************************************
.\"
.\" START OF THE TAB FIELD MACRO
.\"
.\" The .Tf macro moves over to the specified horrizontal position on the line
.\"  If the current input is past the requested position, then a break is
.\"  inserted followed by the movement to the requested position.
.\"
.\" NOTE: the positioning does not work correctly for troff; I believe the
.\"       problem is the length of the diversion is different than the length
.\"       of the string when output.  The results are close, but not perfectly
.\"       aligned
.\"
.de Tf  \" tab field
.\"		Check argument
.if \\n(.$<1 .ab Tf macro requires an argument
.\"		Find the length of the current line: use a diversion to get it
.di }s
.br
.di
.\"		Find length of the diversion plus the trailing space that will
.\"		follow it
.di }t
'ti 0           \"	Indent already part of string }s
\\*(}s\c\ 
.br
.di
.\"		Calculate the length of the line (minus the current indent)
.nr }u \\n(dlu-\\n(.iu
.br
.\"		Now put the original input line back out followed by the
.\"		horizontal positioning requested
'ti 0
\\*(}s\c            \"		Again, indent is already part of string }s
.ie \\n(}u-\\$1>0 \{
.br
\\h'\\$1-1m'\c \}
.el \\h'\\$1-\\n(}uu-1m'\c
..
.\"
.\"
.\"
.ad l
.TH ovlfsmount 8 "Sun Jul 27 22:21:37 EDT 2003" "Arthur Naseef" "Linux System Admin"
.\"
.\" SECTION BREAK
.\"
.SH NAME
ovlfsmount -
mount program for the Overlay Filesystem.
.\"
.\" SECTION BREAK
.\"
.SH VERSION
This manual page covers the version of the
ovlfsmount program which corresponds to
version 2.0.1 of the overlay filesystem.
.\"
.\" SECTION BREAK
.\"
.SH SYNOPSIS
ovlfsmount
.HI
[\-MmnPpsuvxz]
[\-i\ tab\-file]
[\-K\ selector\-name]
[\-o\ mount\-prog]
[\-S\ selector\-name]
[\-U\ mtab\-file]
[fsname ...]
.\"
.\" SECTION BREAK
.\"
.SH DESCRIPTION
The Overlay Filesystem, ovlfs,
takes many mount options in order to
mount even the simplest filesystem configuration,
and the filesystem has
many options available.
When using the standard mount program,
the command line,
or the similar fstab entry,
becomes rather long and difficult to read.
.\"
.P
With the \fBovlfsmount\fR program,
the mount options for the ovlfs
are more easily maintained.
Furthermore, the ovlfsmount program
has the following special features:
.P
.TP
\fBPrefiltering\fR
The contents
of the input file
are filtered
with a command,
which is specified
in the input file itself,
before being interpreted
by ovlfsmount.
So, for example,
it is possible
to use a preprocessor, such as m4,
to make the file more readable and
easier to maintain.
.TP
\fBOption Selection\fR
The filesystem to mount,
as well as
mount options,
are selected
based on the
results of a command.
.\"
.\" SUB-SECTION BREAK
.\"
.SS Options
The following options
are accepted by ovlfsmount
on the command line.
.\"
.TP
.BI \-i	tab\-file
Path to the ovlfs mount-tab file
describing the mounts.
.TP
.BI \-K	selector\-name
Determine the current
selector key
for the named selector.
See the section,
\fBSELECTORS\fR, below
for more information.
Note that no filesystem
is mounted.
.TP
.B \-M
Directly mount the filesystem
using the mount system call
instead of calling
the \fBmount(8)\fR program.
.TP
.B \-m
Use the \fBmount(8)\fR program
instead of mounting
the filesystem directly
with the mount system call.
.TP
.B \-n
Don't actually mount the filesystem,
but show the command
that would be executed instead.
Also displays the media's selector key
found when a selector is used.
.TP
.BI \-o	mount\-prog
Use the specified mount program
to mount the filesystem
instead of
the default mount program
or the mount system call.
.TP
.B \-P
Read the mount table file
and display the list
of mount entries
and selector entries.
Filesystems are not mounted.
.TP
.B \-p
Preprocess the mount table file
and display the result
on standard output.
Note that filesystems
are not mounted.
.TP
.BI \-S	selector\-name
Mount the filesystem
matched by the selector
with the name,
\fBselector\-name\fR.
See the section,
\fBSELECTORS\fR,
below for more information.
.TP
.B \-s
Read the mount table
from standard input
instead of a file.
.TP
.BI \-U	mtab\-file
Update the specified mtab file
with the filesystem information
for successful mounts.
Note that this option
has no effect
when a mount program
is used instead of
the mount system call.
.TP
.B \-u
Update the mtab file
with the filesystem information
for successful mounts.
The mtab file
is not updated
by default.
Note that this option
has no effect
when a mount program
is used instead of
the mount system call.
.TP
.B \-v
Verbose: display the
mount and selector
commands used,
in addition
to the media's selector key,
when a selector is used.
.TP
.B \-x
Display the
mount and selector
commands used.
.TP
.B \-z
Do not update the mtab file.
This is the default.
.\"
.TE
.\"
.\" SECTION BREAK
.\"
.SH USAGE
Once an \fBovlfstab(5)\fR
file is created,
ovlfs filesystem can
be mounted
by name or selector.
To mount a
filesystem by name,
execute ovlfsmount
as follows:
.P
.ti +1i
ovlfsmount \fBfs\-name\fR
.P
Replace \fBfs\-name\fR with
the name of the filesystem,
as it is defined
in the ovlfstab file.
To mount a filesystem
by a selector,
execute ovlfsmount
as follows:
.P
.ti +1i
ovlfsmount -S \fBsel\-name\fR
.P
Replace \fBsel\-name\fR with
the name of the selector.
.\"
.\" SECTION BREAK
.\"
.SH SELECTORS
Selectors allow
the name
of the filesystem
that will be mounted
to be chosen
based on the
output of a command.
Different mount points,
storage files,
and other options
may be set
in this manner.
An example use
of this feature
is the use
of different storage files
and storage roots
for each of many CDs,
based on the identification
for each CD.
.\"
.P
The key to this function
is the selector's
key generation command.
For example,
"cdinfo | md5sum",
could be a good choice
for identifying CDs.
Output from
the selector command
is compared directly
with the key values
stored in the ovlfstab file.
The matching entry
is used to
perform the mount.
When no entry matches,
the program exits
with an error message.
.\"
.P
Two types of selectors
are available: CD and special.
Both operate
in the same manner,
but CD selectors use
a pre-defined key generation command
while special selectors
require the command
to be specified.
Commands specified
in ovlfstab
are not processed
by the shell.
Each argument specified
is passed as-is
to execv(3).
If a shell command
is desired,
use "sh -c"
followed by the
shell command
as a single argument.
For example:
.\"
.P
.ti 1i
Command "/bin/sh" "-c" "echo entire-shell-command-here"
.\"
.SS Configuration
Selectors are defined
in the ovfstab file.
The following information
is required
for each selector:
.BL
.sp 1
.LI
Selector name
used to identify
the selector.
.sp 1
.LI
List of keys
with a filesystem
name for each key.
.sp 1
.LI
Selector command name and arguments
(special selectors only).
.LE
.\"
.P
In addition,
mount options
may be specified
for each key
which will override
the defaults
for the named filesystem.
.\" .\"
.\" .\" SECTION BREAK
.\" .\"
.\" .SH WARNINGS
.\" WARNING INFO HERE
.\"
.\" SECTION BREAK
.\"
.SH PATHS AND COMMANDS
.TP
/etc/ovlfstab
Default path
to the ovlfs
mount-tab file.
.TP
/etc/mtab
Default path
to the mtab file.
.TP
/sbin/mount
Default path
for the mount
program when using
an external mount program.
.TP
"/bin/sh -c 'set -- $(cdinfo | ( md5sum || cksum )); echo $1'"
Default CD selector
key generation command.
Note the lack
of paths for the
cdinfo, m5dsum, and cksum
commands.
.\"
.\" SECTION BREAK
.\"
.SH NOTES
.BL
.LI
There is no unmount
version of this program.
Just use the system
\fBumount(8)\fR program
instead.
Of course,
contributions are welcome.
.LE
.\"
.\" SECTION BREAK
.\"
.SH SEE ALSO
execv(3), mount(8), ovl(8), ovlfstab(5)
.\"
.\" SECTION BREAK
.\"
.SH COPYRIGHT
This manual page and associated source code is licensed under the
GNU General Public License Version 2.
Of course, all source derived from the linux kernel source is copyrighted
by the original author of that source.
.sp 1
COPYRIGHT (C) 2003 By Arthur Naseef
.\"
.\" SECTION BREAK
.\"
.SH FILE VERSION
 @(#)ovlmount.m	1.2	[03/07/27 22:20:33]


ovlfs-2.0.1/RELEASE_NOTES
                               ==================
                               OVERLAY FILESYSTEM
                                 RELEASE NOTES
                                 VERSION 2.0.0
                               ==================


1. ROBUSTNESS
=============

    Due to the amount of time passed since the initially planned release of
the 2.0.0 version, the following has been affected:

	A. Some of the documentation is not up-to-date.

	B. New files and code have been added with no real impact on the
	   software.

	C. Testing with the latest version of the 2.4 kernel has not been
	   completed.

	D. New features are not documented.

	E. Old warnings and issues may not be applicable.

    In spite of these, this software is working and stable enough for release.
I hope you will find this to be the case.

    Note that support for the ReiserFS as the base or storage filesystem is
not complete; multiple hard links to files, and renamed files in the base or
storage filesystem may not be properly resolved by the Overlay Filesystem.
Otherwise, ReiserFS support is functional.



2. KERNEL VERSIONS
==================

    This version of the Overlay Filesystem is built on, and supports, 2.4
series kernels.  Please note that a limited set of 2.4 kernels have actually
been used, so there may be incompatibilities with others due to changes in the
VFS or other portions of the Linux kernel.

    While support for 2.2 kernels was considered, it is not complete; the 2.2
and 2.4 kernels are very similar, but 2.2 lacks many useful interfaces on which
the Overlay Filesystem relies.  Therefore, 2.2 support is currently incomplete.

    Support for 2.0 kernels was present in the 1.0 series for the Overlay
Filesystem, but that support has been lost in the migration to 2.4; in
addition, the 1.0 ovlfs versions had serious flaws that have been corrected.



3. BUG FIXES
============

    Since 2.0.0 is a major rewrite from version 1, listing bug fixes is not
effective.  However, there were serious flaws in version 1 that have been
corrected.  These are listed here:

	A. Directory hierarchy creation used a recursive method that could
	   easily overflow the kernel stack.

	B. Large stack variables were allocated by the ovlfs, which could
	   easily overflow the kernel stack.  All such variables have been
	   replaced with heap memory.

	C. A storage filesystem root that could be reached under the base
	   filesystem root was not supported, and such an instance lead to
	   certain, serious problems in the kernel.  This is no longer true.



4. IMPROVEMENTS
===============

    Many improvements have been made to the Overlay Filesystem.  The majority
support future development efforts and improve the stability and operation of
the filesystem.

    The most notable improvement is the Storage System support which allows
alternative methods of storing Overlay Filesystem data to be added to the core
without changing the core.  Support for storing the filesystem's data on block
devices and in a block-formatted file are planned, which will have a great
impact on the filesystem's required memory use.

    Another notable improvement is the ovlmount program included in this
release.  It simplifies the management of multiple Overlay Filesystem mounts
by using an fstab-like file specifically designed for the Overlay Filesystem.



5. SMP SUPPORT
==============

    At this time, SMP support is not tested, and is expected NOT to work.
The effort needed to find and resolve SMP issues has not been made yet.



6. DEVELOPER'S NOTE
===================

    Much remains to be cleaned up for "correct" kernel coding.  This holds
especially true in the case of module versions.  The current plan for multiple
kernel version support involves splitting the core into different directories,
with one for each major implementation based on the kernel version.  With this
model, the existing code can be greatly reduced in size and readability.

    Any help in this matter is greatly appreciated.



7. LIVE CD CREATED
==================

    With this version of the Overlay Filesystem, I have successfully created a
"live", bootable CD using the Overlay Filesystem as the root filesystem.  Note
that the real root is replaced after kernel startup using a small application
that I wrote.

    The CD boots directly to a SuSE 7.3 system with web browser and all types
of wonderful features.  If you would like to know more, please write!

    One of the great benefits of the live CD is the thorough testing it gives
the Overlay Filesystem!



8. LEX AND YACC
===============

    The mount directory uses lex and yacc (or flex and bison) to generate a
parser for the ovlfstab file.  Output from these programs is included in the
release by default, and will not be removed by "make clean", so that lex and
yacc should not be required on your system.  Note, though, that the flex
library is needed.

    If you prefer to rebuild the lex and yacc output, please note that the
makefiles expect programs named lex and yacc by default.



9. THANKS
=========

    Many thanks goes out to the User Mode Linux project hosted on Sourceforge
at the following URL:

	http://user-mode-linux.sourceforge.net/

Debugging the Overlay Filesystem has been made significantly easier, and
corruption resulting from bugs was limited to a "safe area" (i.e. the content
of a single file).

    Keep up the good work!



[version @(#) RELEASE_NOTES 1.3]


ovlfs-2.0.1/doc/ovlmount.m
.\"
.\" COPYRIGHT (C) 2003 By Arthur Naseef
.\"
.\" This file is covered under the GNU General Public License Version 2.  For
.\"  more information, see the file COPYING.
.\"
.\"
.\" FILE: ovlfsmount.m
.\"
.\" DESCRIPTION:
.\"	Manual page for the ovlfsmount program which is designed to manage
.\"	ovlfs mounts more easily than fstab.
.\"
.\"
.\" REVISION HISTORY
.\"
.\" DATE	AUTHOR		DESCRIPTION
.\" ===========	===============	==============================================
.\" 2003-06-01	ARTHUR N.	Initial Release
.\"
.\"
.\" M4 definitions:
include(tmac.asn)dnl
define(m4_concat, $1$2$3$4$5$6$7$8$9)dnl
define(CURRENT_DATE, translit(esyscmd(date), `
', `'))dnl
define(OVL_VERSION, translit(esyscmd(awk '{print $1; exit}' Version), `
', `'))dnl
.\"
.\"
.\"
.ad l
.TH ovlfsmount 8 "CURRENT_DATE" "Arthur Naseef" "Linux System Admin"
.\"
.\" SECTION BREAK
.\"
.SH NAME
ovlfsmount -
mount program for the Overlay Filesystem.
.\"
.\" SECTION BREAK
.\"
.SH VERSION
This manual page covers the version of the
ovlfsmount program which corresponds to
version OVL_VERSION of the overlay filesystem.
.\"
.\" SECTION BREAK
.\"
.SH SYNOPSIS
ovlfsmount
.HI
[\-MmnPpsuvxz]
[\-i\ tab\-file]
[\-K\ selector\-name]
[\-o\ mount\-prog]
[\-S\ selector\-name]
[\-U\ mtab\-file]
[fsname ...]
.\"
.\" SECTION BREAK
.\"
.SH DESCRIPTION
The Overlay Filesystem, ovlfs,
takes many mount options in order to
mount even the simplest filesystem configuration,
and the filesystem has
many options available.
When using the standard mount program,
the command line,
or the similar fstab entry,
becomes rather long and difficult to read.
.\"
.P
With the \fBovlfsmount\fR program,
the mount options for the ovlfs
are more easily maintained.
Furthermore, the ovlfsmount program
has the following special features:
.P
.TP
\fBPrefiltering\fR
The contents
of the input file
are filtered
with a command,
which is specified
in the input file itself,
before being interpreted
by ovlfsmount.
So, for example,
it is possible
to use a preprocessor, such as m4,
to make the file more readable and
easier to maintain.
.TP
\fBOption Selection\fR
The filesystem to mount,
as well as
mount options,
are selected
based on the
results of a command.
.\"
.\" SUB-SECTION BREAK
.\"
.SS Options
The following options
are accepted by ovlfsmount
on the command line.
.\"
.TP
.BI \-i	tab\-file
Path to the ovlfs mount-tab file
describing the mounts.
.TP
.BI \-K	selector\-name
Determine the current
selector key
for the named selector.
See the section,
\fBSELECTORS\fR, below
for more information.
Note that no filesystem
is mounted.
.TP
.B \-M
Directly mount the filesystem
using the mount system call
instead of calling
the \fBmount(8)\fR program.
.TP
.B \-m
Use the \fBmount(8)\fR program
instead of mounting
the filesystem directly
with the mount system call.
.TP
.B \-n
Don't actually mount the filesystem,
but show the command
that would be executed instead.
Also displays the media's selector key
found when a selector is used.
.TP
.BI \-o	mount\-prog
Use the specified mount program
to mount the filesystem
instead of
the default mount program
or the mount system call.
.TP
.B \-P
Read the mount table file
and display the list
of mount entries
and selector entries.
Filesystems are not mounted.
.TP
.B \-p
Preprocess the mount table file
and display the result
on standard output.
Note that filesystems
are not mounted.
.TP
.BI \-S	selector\-name
Mount the filesystem
matched by the selector
with the name,
\fBselector\-name\fR.
See the section,
\fBSELECTORS\fR,
below for more information.
.TP
.B \-s
Read the mount table
from standard input
instead of a file.
.TP
.BI \-U	mtab\-file
Update the specified mtab file
with the filesystem information
for successful mounts.
Note that this option
has no effect
when a mount program
is used instead of
the mount system call.
.TP
.B \-u
Update the mtab file
with the filesystem information
for successful mounts.
The mtab file
is not updated
by default.
Note that this option
has no effect
when a mount program
is used instead of
the mount system call.
.TP
.B \-v
Verbose: display the
mount and selector
commands used,
in addition
to the media's selector key,
when a selector is used.
.TP
.B \-x
Display the
mount and selector
commands used.
.TP
.B \-z
Do not update the mtab file.
This is the default.
.\"
.TE
.\"
.\" SECTION BREAK
.\"
.SH USAGE
Once an \fBovlfstab(5)\fR
file is created,
ovlfs filesystem can
be mounted
by name or selector.
To mount a
filesystem by name,
execute ovlfsmount
as follows:
.P
.ti +1i
ovlfsmount \fBfs\-name\fR
.P
Replace \fBfs\-name\fR with
the name of the filesystem,
as it is defined
in the ovlfstab file.
To mount a filesystem
by a selector,
execute ovlfsmount
as follows:
.P
.ti +1i
ovlfsmount -S \fBsel\-name\fR
.P
Replace \fBsel\-name\fR with
the name of the selector.
.\"
.\" SECTION BREAK
.\"
.SH SELECTORS
Selectors allow
the name
of the filesystem
that will be mounted
to be chosen
based on the
output of a command.
Different mount points,
storage files,
and other options
may be set
in this manner.
An example use
of this feature
is the use
of different storage files
and storage roots
for each of many CDs,
based on the identification
for each CD.
.\"
.P
The key to this function
is the selector's
key generation command.
For example,
"cdinfo | md5sum",
could be a good choice
for identifying CDs.
Output from
the selector command
is compared directly
with the key values
stored in the ovlfstab file.
The matching entry
is used to
perform the mount.
When no entry matches,
the program exits
with an error message.
.\"
.P
Two types of selectors
are available: CD and special.
Both operate
in the same manner,
but CD selectors use
a pre-defined key generation command
while special selectors
require the command
to be specified.
Commands specified
in ovlfstab
are not processed
by the shell.
Each argument specified
is passed as-is
to execv(3).
If a shell command
is desired,
use "sh -c"
followed by the
shell command
as a single argument.
For example:
.\"
.P
.ti 1i
Command "/bin/sh" "-c" "echo entire-shell-command-here"
.\"
.SS Configuration
Selectors are defined
in the ovfstab file.
The following information
is required
for each selector:
.BL
.sp 1
.LI
Selector name
used to identify
the selector.
.sp 1
.LI
List of keys
with a filesystem
name for each key.
.sp 1
.LI
Selector command name and arguments
(special selectors only).
.LE
.\"
.P
In addition,
mount options
may be specified
for each key
which will override
the defaults
for the named filesystem.
.\" .\"
.\" .\" SECTION BREAK
.\" .\"
.\" .SH WARNINGS
.\" WARNING INFO HERE
.\"
.\" SECTION BREAK
.\"
.SH PATHS AND COMMANDS
.TP
/etc/ovlfstab
Default path
to the ovlfs
mount-tab file.
.TP
/etc/mtab
Default path
to the mtab file.
.TP
/sbin/mount
Default path
for the mount
program when using
an external mount program.
.TP
"/bin/sh -c 'set -- $(cdinfo | ( md5sum || cksum )); echo $1'"
Default CD selector
key generation command.
Note the lack
of paths for the
cdinfo, m5dsum, and cksum
commands.
.\"
.\" SECTION BREAK
.\"
.SH NOTES
.BL
.LI
There is no unmount
version of this program.
Just use the system
\fBumount(8)\fR program
instead.
Of course,
contributions are welcome.
.LE
.\"
.\" SECTION BREAK
.\"
.SH SEE ALSO
execv(3), mount(8), ovl(8), ovlfstab(5)
.\"
.\" SECTION BREAK
.\"
.SH COPYRIGHT
This manual page and associated source code is licensed under the
GNU General Public License Version 2.
Of course, all source derived from the linux kernel source is copyrighted
by the original author of that source.
.sp 1
COPYRIGHT (C) 2003 By Arthur Naseef
.\"
.\" SECTION BREAK
.\"
.SH FILE VERSION
 @(#)ovlmount.m	1.2	[03/07/27 22:20:33]
